### PR TITLE
Repo listing

### PR DESCRIPTION
**Pull request by moliholy**
Related to https://github.com/cvmfs/python-cvmfsutils/pull/5

### CODE REVIEW

File: cvmfs/repository.py [hunk 1]
```diff
@@ -293,6 +293,86 @@ def _retrieve_and_open_catalog(self, catalog_hash):
...
+                break
+            best_fit = best_nested.retrieve_from(self)
+            result = best_fit.find_directory_entry(path)
+        return result
+
+    def lookup(self, path):
+        """
+        Lookups up in the repository for a given path
+        :param path: path to search for
+        follow it and return its final representation
```
Comment (reneme) [inline cvmfs/repository.py:319]: Is that a left over of the symlink resolving?

**Comment by moliholy** [comment]
Closed because it is included in https://github.com/cvmfs/python-cvmfsutils/pull/8

File: cvmfs/repository.py [hunk 2]
```diff
@@ -293,6 +293,86 @@ def _retrieve_and_open_catalog(self, catalog_hash):
...
+        :return: the DirectoryEntry that corresponds to the given path
+        """
+        result = None
+        index = -1
+        if not path or path == "":
```
Comment (reneme) [inline cvmfs/repository.py:324]: the empty string resolves to `false`.

``` python
if not path:
  pass
```

should be enough

File: cvmfs/repository.py [hunk 3]
```diff
@@ -293,6 +293,86 @@ def _retrieve_and_open_catalog(self, catalog_hash):
...
+            path = "/"
+        while index < len(path):
+            index = path.find('/', index + 1)
+            if index < 0:
+                index = len(path)
+                current_path = path
+            else:
+                current_path = path[:index]
+            result = self._lookup_path(current_path)
```
Comment (reneme) [inline cvmfs/repository.py:333]: this does lookup all directory entries on the way, doesn't it? Why is that necessary?

File: cvmfs/repository.py [hunk 4]
```diff
@@ -293,6 +293,86 @@ def _retrieve_and_open_catalog(self, catalog_hash):
...
+            if result is None:
+                return None
+        return result
```
Comment (reneme) [inline cvmfs/repository.py:336]:
``` python
if result is None:
  return None
return result
```

why not simply `return result`?

Comment (moliholy) [inline cvmfs/repository.py:336]: Because they have different indentation. If is None return it, but otherwise keep iterating in the while loop

File: cvmfs/repository.py [hunk 5]
```diff
@@ -293,6 +293,86 @@ def _retrieve_and_open_catalog(self, catalog_hash):
...
+        :param path: path of the directory
+        :return: a list of DirectoryEntry representing all the entries for the
+        given directory, or None if such a directory does not exist
+        """
+        dirent = self.lookup(path)
+        if dirent is not None and dirent.is_directory():
+            if not dirent.is_nested_catalog_mountpoint():
+                best_fit = self._opened_catalogs_for_path(path)
+            else:
+                best_fit = self.retrieve_catalog_for_path(path)
```
Comment (reneme) [inline cvmfs/repository.py:366]: Above is somewhat code duplication to retrieve necessary catalogs (cf. `_lookup_path()`). Along with the introduction of a `CatalogManager` I'd like to have one central place for handling the catalog loading.

Something like:

``` python
def _retrieve_nested_catalogs_for_path(self, path, mount_leaf):
  """ makes sure the catalog responsible for 'path' is loaded.
      if 'mount_leaf' is true and the directory entry is a nested catalog
      mountpoint, it will download the respective nested catalog as well
  """
  pass
```

File: cvmfs/test/repository_test.py [hunk 6]
```diff
@@ -95,3 +95,34 @@ def test_expired_whitelist(self):
...
+        self.mock_repo.make_valid_whitelist()
+        self.mock_repo.serve_via_http()
+        repo = cvmfs.open_repository(self.mock_repo.url,
+                                     public_key=self.mock_repo.public_key)
+        dirent = repo.lookup('/.cvmfsdirtab')
+        self.assertIsNotNone(dirent)
+        dirent = repo.lookup('/bar/4/foo')
+        self.assertIsNotNone(dirent)
+        dirent = repo.lookup('/bar/4/foobar')
+        self.assertIsNone(dirent)
```
Comment (reneme) [inline cvmfs/test/repository_test.py:110]: ~~A check for a non-existing directory entry would be nice.~~ What about relative paths? Paths with a trailing slash, maybe?

File: cvmfs/test/repository_test.py [hunk 7]
```diff
@@ -95,3 +95,34 @@ def test_expired_whitelist(self):
...
+
+    def test_list(self):
+        self.mock_repo.make_valid_whitelist()
+        self.mock_repo.serve_via_http()
+        repo = cvmfs.open_repository(self.mock_repo.url,
+                                     public_key=self.mock_repo.public_key)
+        dirents = repo.list_directory('/')
+        self.assertIsNotNone(dirents)
+        self.assertEqual(3, len(dirents))
+        dirents = repo.list_directory('/bar/3')
```
Comment (reneme) [inline cvmfs/test/repository_test.py:121]: Please also check if the expected directory entries have been listed

File: cvmfs/test/repository_test.py [hunk 8]
```diff
@@ -95,3 +95,34 @@ def test_expired_whitelist(self):
...
+        self.assertIsNotNone(dirents)
+        self.assertEqual(4, len(dirents))
+        dirents = repo.list_directory('/bar/4/foo')
+        self.assertIsNone(dirents)
+        dirents = repo.list_directory('/fakedir')
+        self.assertIsNone(dirents)
```
Comment (reneme) [inline cvmfs/test/repository_test.py:127]: Do those tests involve loading of nested catalogs?

Comment (moliholy) [inline cvmfs/test/repository_test.py:127]: Yes, they do. Only one level, though (because there are no more levels at the moment)

File: cvmfs/repository.py (hunk 4 above)
Comment (reneme) [inline cvmfs/repository.py:336]: d'accord. What about this as a compromise?

``` python
while True:
  # ...
  if result is None:
    break
return result
```

File: cvmfs/repository.py (hunk 5 above)
Comment (moliholy) [inline cvmfs/repository.py:366]: I have just changed that. The lookup operation itself manages the download, so if there is actually a directory entry there then its catalog is also loaded, so there is no need to download it

File: cvmfs/test/repository_test.py (hunk 7 above)
Comment (moliholy) [inline cvmfs/test/repository_test.py:121]: Added in the tests now

File: cvmfs/repository.py (hunk 3 above)
Comment (moliholy) [inline cvmfs/repository.py:333]: Nop, I'll change that

**Comment by moliholy** [comment]
Changes applied here: https://github.com/cvmfs/python-cvmfsutils/pull/8

File: cvmfs/repository.py (hunk 4 above)
Comment (moliholy) [inline cvmfs/repository.py:336]: If you check out the new changes this function is now one-line long ;)
Call it the magic of abstraction
